go/internal/feast/onlinestore/cassandraonlinestore.go

```diff
@@ -0,0 +1,392 @@
+			if iter.NumRows() == 0 {
+				for _, featName := range featureNames {
+					results[rowIdx][featureNamesToIdx[featName]] = FeatureData{
+						Reference: serving.FeatureReferenceV2{
+							FeatureViewName: featureViewName,
+							FeatureName:     featName,
+						},
+						Value: types.Value{
+							Val: &types.Value_NullVal{
+								NullVal: types.Null_NULL,
+							},
+						},
+					}
+				}
+				return
+			}
+
+			scanner := iter.Scanner()
+			var entityKey string
+			var featureName string
+			var eventTs time.Time
+			var valueStr []byte
+			var deserializedValue types.Value
+			rowFeatures := make(map[string]FeatureData)
+			for scanner.Next() {
+				err := scanner.Scan(&entityKey, &featureName, &eventTs, &valueStr)
+				if err != nil {
+					errorsChannel <- errors.New("could not read row in query for (entity key, feature name, value, event ts)")
+					return
+				}
+				if err := proto.Unmarshal(valueStr, &deserializedValue); err != nil {
+					errorsChannel <- errors.New("error converting parsed Cassandra Value to types.Value")
+					return
+				}
+
+				if deserializedValue.Val != nil {
+					// Convert the value to a FeatureData struct
+					rowFeatures[featureName] = FeatureData{
+						Reference: serving.FeatureReferenceV2{
+							FeatureViewName: featureViewName,
+							FeatureName:     featureName,
+						},
+						Timestamp: timestamppb.Timestamp{Seconds: eventTs.Unix(), Nanos: int32(eventTs.Nanosecond())},
+						Value: types.Value{
+							Val: deserializedValue.Val,
+						},
+					}
+				} else {
+					// Return FeatureData with a null value
+					rowFeatures[featureName] = FeatureData{
+						Reference: serving.FeatureReferenceV2{
+							FeatureViewName: featureViewName,
+							FeatureName:     featureName,
+						},
+						Value: types.Value{
+							Val: &types.Value_NullVal{
+								NullVal: types.Null_NULL,
+							},
+						},
+					}
+				}
+			}
+
+			if err := scanner.Err(); err != nil {
+				errorsChannel <- errors.New("failed to scan features: " + err.Error())
+				return
+			}
+
+			for _, featName := range featureNames {
+				featureData, ok := rowFeatures[featName]
+				if !ok {
+					featureData = FeatureData{
+						Reference: serving.FeatureReferenceV2{
+							FeatureViewName: featureViewName,
+							FeatureName:     featName,
+						},
+						Value: types.Value{
+							Val: &types.Value_NullVal{
+								NullVal: types.Null_NULL,
+							},
+						},
+					}
+				}
+				results[rowIdx][featureNamesToIdx[featName]] = featureData
+			}
+		}(serializedEntityKey)
+	}
+
+	// wait until all concurrent single-key queries are done
+	waitGroup.Wait()
+	close(errorsChannel)
+
+	var collectedErrors []error
+	for err := range errorsChannel {
+		if err != nil {
+			collectedErrors = append(collectedErrors, err)
+		}
+	}
+	if len(collectedErrors) > 0 {
+		return nil, errors.Join(collectedErrors...)
+	}
+
+	return results, nil
+}
+
+func (c *CassandraOnlineStore) Destruct() {
```

Comment:
What does this Destruct() do? Maybe we should close the session here if my understanding is correct?
I see you are opening a session with "CreateTracedSession"...

Reply:
Was following the lead of `redisonlinestore.go` with empty `Destruct()`. TBH I guess it won't matter a whole lot, as when `Destruct()` needs to be called, the container is getting killed anyway, maybe the db will have a dead connection for a while.

Anyway, closing the session there now, thanks.